front/lib/auth.ts

```diff
@@ -797,21 +811,90 @@ export class Authenticator {
     return this._groups.map((g) => g.toJSON());
   }
 
-  hasPermission(acls: ACLType[], permission: Permission): boolean {
-    // For each acl, does the user belongs to a group that has the permission?
-    return acls.every((acl) =>
-      this.groups().some((group) =>
-        groupHasPermission(acl, permission, group.id)
+  /**
+   * Checks if the user has the specified permission across all resource permissions.
+   *
+   * This method applies a conjunction (AND) over all resource permission entries. The user
+   * must have the required permission in EVERY entry for the check to pass.
+   */
+  hasPermission(
+    resourcePermissions: ResourcePermission[],
+    permission: PermissionType
+  ): boolean {
+    // Apply conjunction (AND) over all resource permission entries.
+    return resourcePermissions.every((rp) =>
+      this.hasResourcePermission(rp, permission)
+    );
+  }
+
+  /**
+   * Determines if a user has a specific permission on a resource based on their role and group
+   * memberships.
+   *
+   * The permission check follows two independent paths (OR):
+   *
+   * 1. Role-based permission check:
+   *    Applies when the resource has role-based permissions configured.
+   *    Permission is granted if:
+   *    - The resource has public access (role="none") for the requested permission, OR
+   *    - The user's role has the required permission AND the resource belongs to user's workspace
+   *
+   * 2. Group-based permission check:
+   *    Applies when the resource has group-based permissions configured.
+   *    Permission is granted if:
+   *    - The user belongs to a group that has the required permission on this resource
+   *
+   * @param resourcePermission - The resource's permission configuration
+   * @param permission - The specific permission being checked
+   * @returns true if either permission path grants access
+   */
+  private hasResourcePermission(
+    resourcePermission: ResourcePermission,
+    permission: PermissionType
+  ): boolean {
+    // First path: Role-based permission check.
+    if (hasRolePermissions(resourcePermission)) {
+      const workspace = this.getNonNullableWorkspace();
+
+      // Check for public access first.
```

Comment:
Let's add to the comment that this is the only case of cross workspace permissioning.